packages/flutter/lib/src/rendering/viewport.dart

```diff
@@ -66,13 +68,72 @@ class RenderViewport extends RenderBox with RenderObjectWithChildMixin<RenderBox
   Axis get scrollDirection => _scrollDirection;
   Axis _scrollDirection;
   void set scrollDirection(Axis value) {
+    assert(value != null);
     if (value == _scrollDirection)
       return;
-    assert(_offsetIsSane(scrollOffset, value));
+    assert(_offsetIsSane(_paintOffset, value));
     _scrollDirection = value;
     markNeedsLayout();
   }
 
+  Painter get overlayPainter => _overlayPainter;
+  Painter _overlayPainter;
+  void set overlayPainter(Painter value) {
+    if (_overlayPainter == value)
+      return;
+    if (attached)
+      _overlayPainter?.detach();
+    _overlayPainter = value;
+    if (attached)
+      _overlayPainter?.attach(this);
+    markNeedsPaint();
+  }
+
+  void attach() {
+    super.attach();
+    _overlayPainter?.attach(this);
+  }
+
+  void detach() {
+    super.detach();
+    _overlayPainter?.detach();
+  }
+
+  Offset get _paintOffsetRoundedToIntegerDevicePixels {
+    final double devicePixelRatio = ui.window.devicePixelRatio;
+    int dxInDevicePixels = (_paintOffset.dx * devicePixelRatio).round();
+    int dyInDevicePixels = (_paintOffset.dy * devicePixelRatio).round();
+    return new Offset(dxInDevicePixels / devicePixelRatio,
+                      dyInDevicePixels / devicePixelRatio);
+  }
+
+  void applyPaintTransform(RenderBox child, Matrix4 transform) {
+    final Offset effectivePaintOffset = _paintOffsetRoundedToIntegerDevicePixels;
+    super.applyPaintTransform(child, transform.translate(effectivePaintOffset.dx, effectivePaintOffset.dy));
+  }
+
+}
+
+/// A render object that's bigger on the inside.
+///
+/// The child of a viewport can layout to a larger size than the viewport
+/// itself. If that happens, only a portion of the child will be visible through
+/// the viewport. The portion of the child that is visible is controlled by the
+/// scroll offset.
+///
+/// Viewport is the core scrolling primitive in the system, but it can be used
```

Comment:
Isn't Scrollable the core scrolling primitive? I guess Viewport is just a clipping building block for Scrollables and others.